### ISSUE TITLE
Capturing Multiple Returns

### ISSUE THREAD

**Issue by cancan101** [blink1073/oct2py#66]
I have a function that looks like:
```
function [f, h1f, h2f, A, theta, psi] = ...
	monofilt(im, nscale, minWaveLength, mult, sigmaOnf, orientWrap)
```
Doing this:
```
ret = %octave monofilt(b, 3, 4, 2, 0.65) -i b
```
yields only the `f` return value. How do I capture the remaining values?

**Comment by blink1073** [comment]
Nargout is inferred by oct2py, so you need to put all of the variables on the lhs.

**Comment by cancan101** [comment]
This does not seem to work either:
```
In [77]: f,h1f,h2f,A,theta,psi = %octave monofilt(b, 3, 4, 2, 0.65) -i b

ans = 
{
  [1,1] =

    -0.41105  -0.13702
    0.13702  0.41105

  [1,2] =

  1.0e-03  *

    -8.45872  -2.81957
    2.81957  8.45872

  [1,3] =

  1.0e-05  *

    -1.30714  -0.43571
    0.43571  1.30714

}
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-77-2bec253f83ef> in <module>()
----> 1 f,h1f,h2f,A,theta,psi = get_ipython().magic(u'octave monofilt(b, 3, 4, 2, 0.65) -i b')

ValueError: need more than 3 values to unpack
```

**Comment by blink1073** [comment]
Whoops, I was thinking from a direct command perspective.  You need to use the `-o` option for the magic.  Try:

```python
%octave monofilt(b, 3, 4, 2, 0.65) -i b -o f,h1f,h2f,A,theta,psi
print(f,h1f,h2f,A,theta,psi)
```

**Comment by cancan101** [comment]
That worked. 
So the first argument is automatically / implicitly copied from Octave to Python but the remaining but be specified?

**Comment by blink1073** [comment]
When you call a method of an Oct2Py object or `eval`, the number of items on the LHS is reflectwd in the call to Octave, but if there aren't any then `ans` is caputured. The magic works differently to be consistent with other magics.

**Comment by cancan101** [comment]
This might be related to that. I am getting different results for these two different calls:
```
In [101]:

octave.monofilt(b, 3, 4, 2, 0.65)
Out[101]:
[array([[ 0.,  0.],
        [ 0.,  0.]]), array([[ 0.,  0.],
        [ 0.,  0.]]), array([[ 0.,  0.],
        [ 0.,  0.]])]
```
vs this:
```
In [102]:

%octave monofilt(b, 3, 4, 2, 0.65) -i b -o f,h1f,h2f,A,theta,psi
ans = 
{
  [1,1] =

    -0.41105  -0.13702
    0.13702  0.41105

  [1,2] =

  1.0e-03  *

    -8.45872  -2.81957
    2.81957  8.45872

  [1,3] =

  1.0e-05  *

    -1.30714  -0.43571
    0.43571  1.30714

}
Out[102]:
[array([[-0.41104991, -0.13701664],
        [ 0.13701664,  0.41104991]]), array([[-0.00845872, -0.00281957],
        [ 0.00281957,  0.00845872]]), array([[ -1.30713648e-05,  -4.35712159e-06],
        [  4.35712159e-06,   1.30713648e-05]])]
```

**Comment by blink1073** [comment]
So in the first case nargout is 1, is that the expect output?  For the second, you need to put a semicolon after the function call to suppress the output since it is `eval`ed.  Otherwise, are those the expected outputs when nargout=6?

**Comment by cancan101** [comment]
Ok. Adding the semicolon to improve readability:
```
In [101]:

octave.monofilt(b, 3, 4, 2, 0.65)
Out[101]:
[array([[ 0.,  0.],
        [ 0.,  0.]]), array([[ 0.,  0.],
        [ 0.,  0.]]), array([[ 0.,  0.],
        [ 0.,  0.]])]
In [113]:

%octave monofilt(b, 3, 4, 2, 0.65); -i b -o f,h1f,h2f,A,theta,psi
Out[113]:
[array([[-0.41104991, -0.13701664],
        [ 0.13701664,  0.41104991]]), array([[-0.00845872, -0.00281957],
        [ 0.00281957,  0.00845872]]), array([[ -1.30713648e-05,  -4.35712159e-06],
        [  4.35712159e-06,   1.30713648e-05]])]
```

All 0s is not the expected output. 

The second case it my reference and I believe the values look good.

**Comment by cancan101** [comment]
Actually on closer inspection, I do not think this is working right:
```
In [28]:

b = np.array([[1, 2], [3, 4]])
%octave monofilt(b, 3, 4, 2, 0.65); -i b -o f,h1f,h2f,A,theta,psi
print(f,h1f,h2f,A,theta,psi)
(None, None, None, None, None, None)
```

**Comment by blink1073** [comment]
I don't think you can define an input and use it in the same block.  Same for output.

**Comment by cancan101** [comment]
Still does not look to work:
```
In [30]:

b = np.array([[1, 2], [3, 4]])
In [31]:

%octave monofilt(b, 3, 4, 2, 0.65); -i b -o f,h1f,h2f,A,theta,psi
Out[31]:
[array([[-0.41104991, -0.13701664],
        [ 0.13701664,  0.41104991]]), array([[-0.00845872, -0.00281957],
        [ 0.00281957,  0.00845872]]), array([[ -1.30713648e-05,  -4.35712159e-06],
        [  4.35712159e-06,   1.30713648e-05]])]
In [32]:

print(f,h1f,h2f,A,theta,psi)
(None, None, None, None, None, None)
```

**Comment by blink1073** [comment]
Hmm, try with no space after `-o`

**Comment by cancan101** [comment]
Nope. Using the `roundtrip` from the docs (with semicolons added to assignments):
```
In [44]:

%octave roundtrip(b); -i b -o q,g
Out[44]:
array([[1, 2],
       [3, 4]])
In [45]:

print(q,g)
(None, None)
In [46]:

%octave roundtrip(b); -i b -oq,g
Out[46]:
array([[1, 2],
       [3, 4]])
In [47]:

print(q,g)
(None, None)
```

**Comment by blink1073** [comment]
Ah, I finally see the problem with the magic:

```python
%octave f,h1f,h2f,A,theta,psi = octave.monofilt(b, 3, 4, 2, 0.65); -i b -o f,h1f,h2f,A,theta,psi 
```
or without using magic:

```python
f,h1f,h2f,A,theta,psi = octave.monofilt(b, 3, 4, 2, 0.65)
print(f,h1f,h2f,A,theta,psi)
```

**Comment by blink1073** [comment]
Whoops, you'll need brackets:

```python
%octave [f,h1f,h2f,A,theta,psi] = octave.monofilt(b, 3, 4, 2, 0.65); -i b -o f,h1f,h2f,A,theta,psi 
```

**Comment by blink1073** [comment]
Time to add a few more examples to the [notebook](http://nbviewer.ipython.org/github/blink1073/oct2py/blob/master/example/octavemagic_extension.ipynb?create=1) methinks...

**Comment by cancan101** [comment]
Should be:
```
%octave [f,h1f,h2f,A,theta,psi] = monofilt(a, 3, 4, 2, 0.65); -i a -o f,h1f,h2f,A,theta,psi
```

**Comment by cancan101** [comment]
The example without the magic does not work:
```
In [79]:

f,h1f,h2f,A,theta,psi = octave.monofilt(b, 3, 4, 2, 0.65)
print(f,h1f,h2f,A,theta,psi)
([array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]])], [array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]])], [array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]])], [array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]])], [array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]])], [array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]]), array([[ 0.,  0.],
       [ 0.,  0.]])])
```

**Comment by blink1073** [comment]
I'm not sure what to say about the non-magic version not working.  It is getting the right number of nargout...   Can you share the `monofilt` function?

**Comment by cancan101** [comment]
http://www.csse.uwa.edu.au/~pk/research/matlabfns/PhaseCongruency/monofilt.m

**Comment by blink1073** [comment]
We've exposed a source of friction.  Oct2Py sends integers to Octave as integer type, since they are interpreted as such by numpy. The following works:

```python
f,h1f,h2f,A,theta,psi = octave.monofilt(b,3.,4.,2.,0.65)
```

**Comment by cancan101** [comment]
Why does the magic version perform differently?

**Comment by blink1073** [comment]
In the magic version, everything is wrapped in octave.eval, so the arguments are sent as plain text, instead of through a MAT file.

—

**Comment by cancan101** [comment]
Okay. If that fact isn't in the docs, it would be worth adding it. I
assumed the two worked using the same mechanism.

**Comment by cancan101** [comment]
Something else is still not right between these two. 
I download this file: http://4.bp.blogspot.com/-ruV5mllFtxA/TjosZziIddI/AAAAAAAAAIw/o8aU8O1ZfqM/s1600/BlackWhite.gif
and then load it using:
```python
import numpy as np
from PIL import Image
a = np.asarray(Image.open('BlackWhite.gif'))/255.
```

This takes a few seconds:
```
%octave [f,h1f,h2f,A,theta,psi] = monofilt(a, 3, 4, 2, 0.65); -i a -o f,h1f,h2f,A,theta,psi
```
but this never returns / takes a VERY long time:
```python
f,h1f,h2f,A,theta,psi = octave.monofilt(a, 3., 4., 2., 0.65)
```

**Comment by blink1073** [comment]
Thanks for the feedback.   The problem was that the output was not suppressed on the second version, causing it to print a *lot* to the console.  I'm going to release version `3.0` that converts `int -> float` by default and suppresses that output (hopefully by this weekend).

**Comment by blink1073** [comment]
I just released 3.0, which fixes this issue.
